### ISSUE TITLE
Add an option to override used nix package

### ISSUE THREAD

**Issue by dundalek** [dundalek/lazy-lsp.nvim#15]
Came up in https://github.com/dundalek/lazy-lsp.nvim/pull/6 . The issue got fixed by making updates to the plugin and I argued that not having the option will incentivize people to report issues which will make the plugin better for everyone.

But in https://github.com/dundalek/lazy-lsp.nvim/issues/13#issuecomment-1458109767 there came up another need due to mismatch when using different nix channels. This is out of the plugin control (besides maintaining different set of configs for different channels which I wouldn't want getting into), so providing an option to override for the user would make sense.

One remaining decision remains what the option should be. The issue is that currently we just pass the options through to the lspcofing. By adding an option we could end up mixing options, which could be confusing or leading to potential conflicts. I can think of two options:

1) Add a key under `configs` for a given server like `nix_pkg`. We then remove the key from table before passing it down to lspconfig. There is a risk options could conflict.
1b. Variant of 1. to increase the chance the option stays unique, we could use even more unique name like `lazy_lsp_nix_pkg`.
2) Add separate option like `packages` which would make lsp configuration and package configuration independent, there would not be any potential for conflict. It is less convenient since the server name will end up duplicated, but it should be in rare situations (e.g. when being on a different nix channel).
```lua
packages = {
  lua_ls = "sumneko-lua-language-server"
}
```

In the trade-off between confusion vs less conveniece (duplicate server name), I am slightly leaning towards 2.

**Comment by wesleyjrz** [comment]
> Came up in #6 . The issue got fixed by making updates to the plugin and I argued that not having the option will incentivize people to report issues which will make the plugin better for everyone.

I strongly agree with you. Most of the time this problem with package names happens because of changes in the nixpkgs repository, but this changes always come to the unstable channel first.

A good option would be to use unstable nix channel as default (as you've said before, it's better for development). I've been thinking about a way to detect this changes automatically, but nothing came to my mind yet.

**Comment by wesleyjrz** [comment]
Btw, there's a nix [module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/command-not-found/command-not-found.nix) called `command-not-found` which is useful to find package names from commands, the only problem is that it returns more than one value sometimes.

![screenshot-20230308-100230](https://user-images.githubusercontent.com/60184588/223720786-f0b4ef68-f108-44cc-83d1-183d35807971.png)

![screenshot-20230308-100320](https://user-images.githubusercontent.com/60184588/223720828-ff363e39-8862-4b37-9e72-65e2bbbd517f.png)

**Comment by zoriya** [comment]
I found another use case for this. The C# LSP, omnisharp, needs to run on exactly the same SDK as your project and the current definition is:
```lua
omnisharp = {
    cmd = {
      "OmniSharp"
    },
    pkgs = {
      "omnisharp-roslyn",
      "dotnet-sdk"
    }
  },
```
Which hard codes the last LTS SDK version (version 6 as of now).

I personally would like to remove the `dotnet-sdk` from the pkgs array to allow omnisharp to pick the SDK from the environment, but this can't be the default in lazy-lsp since it simply breaks the LSP if no SDK is available on the path.

**Comment by dundalek** [comment]
@zoriya Thanks for sharing, this use case makes a lot of sense.

I imagine the sdk versions could differ between projects, so we might need to pass a function to allow the flexibility to handle it properly.

Given that, here's what I am thinking:
The user could specify their overrides in `on_new_config` callback and to make it composable we could expose a helper like `in_shell(pkgs, cmd)` that would wrap the command to run in nix-shell with given packages.

```lua
configs = {
  omnisharp = {
    on_new_config = function(new_config, new_root_dir)
      lspconfig.omnisharp.document_config.default_config.on_new_config(new_config, new_root_dir)
      new_config.cmd = lazy_lsp.in_shell({ "omnisharp-roslyn" }, new_config.cmd)
    end,
  },
},
```

That way one could also have per project overrides for example:
```lua
on_new_config = function(new_config, new_root_dir)
  local pkg = "clang-tools"
  if some_project then
    pkg = "clang-tools_16"
  elseif other_project then
    pkg = "clang-tools_15"
  end
  new_config.cmd = lazy_lsp.in_shell({ pkg }, new_config.cmd)
end
```

It will require to change we handle `on_new_config` internally, but it seems to be worth it.

**Comment by zoriya** [comment]
I don't really think if this is that valuable, if the LSP depends on the project I would be more incline in specifying the LSP in a `shell.nix` of the project instead of manually editing the config every time.
The reason why I can't do that right now with `omnisharp` is because even if I specify another `dotnet-sdk`, the LSP has two SDK available and does not know which one to pick. What I want is not to specify another SDK version, but to simply remove it from the packages to includes in the nix-shell.

On another note, I do think that most LSP would benefit running in a `nix-shell --pure` environment to remove other packages or env variable that could affect the LSP unexpectedly, but this should be configurable per LSP in the user's config because sometimes, you do want to edit env var or the path for the LSP (as in my case with `omnisharp`).

Thanks for responding this quickly!

**Comment by dundalek** [comment]
Gotcha, that's a good point with local `shell.nix`.
However, the first example (although bit more verbose that a config option) should still solve your use case since the override does not specify any dotnet-sdk, correct?

The reason I came up with this approach is that I also had another thing in mind. I recently discovered that `jdtls` does not work because of a different binary name. But simple `{ cmd = "jdt-language-server" }` override does not work, because it strips the parameters which that lsp needs.

So having that mechanism internally, that could be solved like:
```lua
new_config.cmd = lazy_lsp.in_shell(pkg, replace_first(new_config.cmd, "jdt-language-server"))
```
